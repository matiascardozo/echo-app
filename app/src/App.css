.App {
    min-height: 100vh;
}

.background-app{
  box-shadow: inset 0 0 5rem rgba(0, 0, 0, .5);

}

.cursor-pointer{
  cursor: pointer;
}

