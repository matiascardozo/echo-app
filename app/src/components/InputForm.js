import React from 'react';

function InputForm({message, onInputChange, onSubmit}) {
    return (
        <form className="form-inline" onSubmit={onSubmit}>
            <div className="form-group">
                <label className="my-1 mr-2"  for="messageInput">Mensaje</label>
                <input
                    type="text"
                    className="form-control-sm my-1 mr-sm-2"
                    id="messageInput"
                    aria-describedby="emailHelp"
                    value={message}
                    onChange={onInputChange}
                />
            </div>
            <button type="submit" className="btn btn-primary btn-sm my-1">
                Submit
            </button>
        </form>
    );
}

export default InputForm;
