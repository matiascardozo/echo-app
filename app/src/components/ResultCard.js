import { faTrash } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import React from 'react';

function ResultCard({ result, clearElement }) {
    return (
        <div class={`card ${result.statusCode ? 'border-danger' : 'border-success'} mb-2 border-3`}>
            <div class="card-body text-dark">
                <div class="row no-gutters">
                    <div class="col"><span>{result.message}</span></div>
                    <div class="col-sm-1">
                        <FontAwesomeIcon onClick={clearElement} className="cursor-pointer" icon={faTrash} />
                    </div>
                </div>
            </div>
        </div>
    );
}

export default ResultCard;
