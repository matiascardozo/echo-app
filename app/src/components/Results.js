import React from 'react';
import ResultCard from './ResultCard';

function Results({ results, clearElement }) {
    return results.map((result, index) => <ResultCard key={index} index={index} result={result} clearElement={clearElement} />);
}

export default Results;
