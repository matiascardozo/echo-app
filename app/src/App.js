import React, { useState } from 'react';
import './App.css';
import InputForm from './components/InputForm';
import axios from 'axios';
import Results from './components/Results';

function App() {
    const [message, setMessage] = useState('');
    const [error, setError] = useState('');
    const [responses, setResponses] = useState([]);
    const [loading, setLoading] = useState(false);

    const onInputChange = (evt) => {
        setMessage(evt.target.value);
    };

    const onSubmit = (evt) => {
        evt.preventDefault();
        setLoading(true);
        axios
            .post('', { message })
            .then((response) => setResponses((prevReponses) => [...prevReponses, response.data]))
            .catch((e) => {
                if (e.response) {
                    setResponses((prevReponses) => [...prevReponses, e.response.data]);
                } else {
                    setError('Unexpected error');
                }
            })
            .finally(() => {
                setMessage('');
                setLoading(false);
            });
    };

    const clearElement = (index) => {
        setResponses((prevReponses) => prevReponses.filter((_, i) => i != index));
    };

    return (
        <div className="container-fluid background-app ">
            <div className="container d-flex w-100 h-100 p-3 mx-auto flex-column align-center App ">
                <header className="masthead mb-auto">
                    <h1 className="text-white">Echo App</h1>
                </header>
                <main role="main" className="inner cover">
                    <div className="row align-items-center">
                        <div className="col-md-6 text-center font-weight-normal">
                            <h4>Type Something</h4>
                            <InputForm
                                message={message}
                                onInputChange={onInputChange}
                                onSubmit={onSubmit}
                                loading={loading}
                            />
                        </div>
                        <div className="col-md-6">
                            {error ? (
                                <span>Unexpected error</span>
                            ) : (
                                <Results results={responses} clearElement={clearElement} />
                            )}
                        </div>
                    </div>
                </main>
                <footer className="mastfoot mt-auto"></footer>
            </div>
        </div>
    );
}

export default App;
