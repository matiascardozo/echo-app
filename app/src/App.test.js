import React from 'react';
import { render } from '@testing-library/react';
import App from './App';
import ResultCard from './components/ResultCard';
import renderer from 'react-test-renderer';
import Results from './components/Results';
import InputForm from './components/InputForm';

test('renders learn react link', () => {
    const { getByText } = render(<App />);
    const linkElement = getByText(/Echo App/i);
    expect(linkElement).toBeInTheDocument();
});

describe('Result card', () => {
    const handleRemoveItem = jest.fn();
    let component;
    const result = { message: 'Hello world' };

    beforeEach(() => {
        component = renderer.create(
            <ResultCard result={result} clearElement={() => handleRemoveItem(1)} />,
        );
    });

    it('renders the message', () => {
        // console.log(component.root.findByType('p'))
        expect(component.root.findByType('span').props.children).toEqual('Hello world');
    });

    it('calls clearElement on icon click', () => {
        component.root.findByType('svg').props.onClick();
        expect(handleRemoveItem).toHaveBeenCalledTimes(1);
        expect(handleRemoveItem).toHaveBeenCalledWith(1);
    });
});

describe('Result List', () => {
    const resultList = [{ message: 'Hello world' }, { message: 'Bye bye!' }];
    it('renders two items', () => {
        const component = renderer.create(<Results results={resultList} />);
        expect(component.root.findAllByType(ResultCard).length).toEqual(2);
    });
});

describe('Intput Form', () => {
    const inputFormProps = {
        message: 'Hello world',
        onInputChange: jest.fn(),
        onSubmit: jest.fn(),
    };
    let component;
    beforeEach(() => {
        component = renderer.create(<InputForm {...inputFormProps} />);
    });
    it('renders the input field with its value', () => {
        const value = component.root.findByType('input').props.value;
        expect(value).toEqual('Hello world');
    });

    it('changes the input field', () => {
        const pseudoEvent = { target: 'Test' };
        component.root.findByType('input').props.onChange(pseudoEvent);
        expect(inputFormProps.onInputChange).toHaveBeenCalledTimes(1);
        expect(inputFormProps.onInputChange).toHaveBeenCalledWith(pseudoEvent);
    });
    it('submits the form', () => {
        const pseudoEvent = {};
        component.root.findByType('form').props.onSubmit(pseudoEvent);
        expect(inputFormProps.onSubmit).toHaveBeenCalledTimes(1);
        expect(inputFormProps.onSubmit).toHaveBeenCalledWith(pseudoEvent);
    });

    it('disables the button and prevents submit', () => {
        component.update(<InputForm {...inputFormProps} loading={true} />);
        expect(component.root.findByType('button').props.disabled).toBeTruthy();
    });

    test('renders snapshot', () => {
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
