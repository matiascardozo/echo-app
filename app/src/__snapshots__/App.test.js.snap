// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Intput Form renders snapshot 1`] = `
<form
  className="form"
  onSubmit={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {},
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
>
  <div
    className="form-group"
  >
    <input
      aria-describedby="emailHelp"
      className="form-control-sm my-1 mr-sm-2"
      id="messageInput"
      onChange={
        [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "target": "Test",
              },
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
      type="text"
      value="Hello world"
    />
  </div>
  <button
    className="btn btn-primary btn-sm my-1"
    type="submit"
  >
    Submit
  </button>
</form>
`;
